File: Solutions/merge_sort.js
export function mergeSort(arr) {
  if (arr.length <= 1) {
    return arr;
  }

  const mid = Math.floor(arr.length / 2);
  const left = arr.slice(0, mid);
  const right = arr.slice(mid);

  const sortedLeft = mergeSort(left);
  const sortedRight = mergeSort(right);

  const result = [];
  let i = 0,
    j = 0;

  while (i < sortedLeft.length && j < sortedRight.length) {
    if (sortedLeft[i] <= sortedRight[j]) {
      result.push(sortedLeft[i]);
      i++;
    } else {
      result.push(sortedRight[j]);
      j++;
    }
  }

  while (i < sortedLeft.length) {
    result.push(sortedLeft[i]);
    i++;
  }

  while (j < sortedRight.length) {
    result.push(sortedRight[j]);
    j++;
  }

  return result;
}

File: 25_1/s_18/exo2/merge_sort.go
package main

import (
	"exo2/sorts"
	"fmt"
)

func mergeSort(items []int) []int {

	if len(items) < 2 {
		return items
	}

	left := mergeSort(items[:len(items)/2])
	right := mergeSort(items[len(items)/2:])

	return merge(left, right)

}

func merge(a []int, b []int) []int {
	final := []int{}
	i := 0
	j := 0

	for i < len(a) && j < len(b) {
		if a[i] < b[j] {
			final = append(final, a[i])
			i++
		} else {
			final = append(final, b[j])
			j++
		}
	}

	for ; i < len(a); i++ {
		final = append(final, a[i])
	}

	for ; j < len(b); j++ {
		final = append(final, b[j])
	}
	return final
}

func main() {
	a := []int{10, 6, 2, 1, 5, 8, 3, 4, 7, 9}
	sorted := sorts.MergeSort(a)
	fmt.Println("Result (MergeSort):", sorted)
}
